File: arena-xunit/tests/Unit/KafkaDependencyBuilderSerializationTest.cs
using System;
using System.Linq;
using ArenaDotnet.Xunit.Dep;
using ArenaDotnet.Xunit.Support;
using Newtonsoft.Json.Linq;
using Xunit;

namespace ArenaDotnet.Xunit.UnitTest;

public class KafkaDependencyBuilderSerializationTest
{
    [Fact]
    public void Build_DefaultPort_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("kafka", obj["type"]);
        Assert.Equal(9092, obj["port"]);
        Assert.NotNull(obj["identifier"]);
    }

    [Fact]
    public void Build_CustomPort_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").WithPort(9192).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal(9192, obj["port"]);
    }

    [Fact]
    public void Build_ApacheNativeFlavor_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").WithFlavor(KafkaFlavor.ApacheNative).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("apache_native", obj["flavor"]);
    }

    [Fact]
    public void Build_ConfluentFlavor_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").WithFlavor(KafkaFlavor.Confluent).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("confluent", obj["flavor"]);
    }

    [Fact]
    public void Build_Identifier_MatchesPattern()
    {
        var dep = new KafkaDependencyBuilder("test").Build();
        Assert.StartsWith("arena-kafka-", dep.Identifier);
    }

    [Fact]
    public void Build_WithImageName_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").WithImageName("apache/kafka").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("apache/kafka", obj["image_name"]);
    }

    [Fact]
    public void Build_WithContainerName_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").WithContainerName("my-kafka").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("my-kafka", obj["container_name"]);
    }

    [Fact]
    public void Build_WithTopic_SerializesCorrectJson()
    {
        var dep = new KafkaDependencyBuilder("test").WithTopic("orders").WithTopic("payments").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var topics = obj["topics"];
        Assert.NotNull(topics);
        Assert.Equal(2, topics.Count());
        Assert.Equal("orders", topics[0]);
        Assert.Equal("payments", topics[1]);
    }

    [Fact]
    public void Build_NoTopics_OmitsTopicsFromJson()
    {
        var dep = new KafkaDependencyBuilder("test").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Null(obj["topics"]);
    }

    [Fact]
    public void Build_ThenMutateBuilderTopics_BuiltDependencyUnaffected()
    {
        var builder = new KafkaDependencyBuilder("test").WithTopic("orders");
        var dep = builder.Build();
        builder.WithTopic("payments");
        Assert.Single(dep.Topics!);
        Assert.Equal("orders", dep.Topics![0]);
    }
}

File: arena-junit/src/test/java/arena/junit/dep/KafkaDependencyBuilderSerializationTest.java
package arena.junit.dep;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import arena.junit.match.ArenaRunnableDependency;
import arena.junit.support.ArenaJson;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import org.junit.jupiter.api.Test;

final class KafkaDependencyBuilderSerializationTest {

  static final class StubDependency implements ArenaRunnableDependency {
    @Override
    public ObjectNode forFfi() {
      return ArenaJson.object().put("identifier", "child");
    }
  }

  @Test
  void build_minimalName_serializesTypeAndEmptyTopics() {
    ObjectNode config = new KafkaDependencyBuilder("kafka").build().forFfi();
    assertEquals("kafka", config.path("type").asText());
    assertTrue(config.path("identifier").asText().startsWith("arena-kafka-kafka-"));
    assertTrue(config.path("topics").isEmpty());
    assertFalse(config.has("children"));
  }

  @Test
  void withTopic_multipleCalls_appendsTopicsArray() {
    ObjectNode config =
        new KafkaDependencyBuilder("kafka").withTopic("orders").withTopic("payments").build().forFfi();
    ArrayNode topics = (ArrayNode) config.path("topics");
    assertEquals(2, topics.size());
    assertEquals("orders", topics.get(0).asText());
    assertEquals("payments", topics.get(1).asText());
  }

  @Test
  void withFlavor_confluent_serializesFlavorValue() {
    ObjectNode config =
        new KafkaDependencyBuilder("kafka").withFlavor(KafkaFlavor.CONFLUENT).build().forFfi();
    assertEquals("confluent", config.path("flavor").asText());
  }

  @Test
  void withFlavor_apacheNative_serializesFlavorValue() {
    ObjectNode config =
        new KafkaDependencyBuilder("kafka").withFlavor(KafkaFlavor.APACHE_NATIVE).build().forFfi();
    assertEquals("apache_native", config.path("flavor").asText());
  }

  @Test
  void withPortAndContainerName_setsScalarFields() {
    ObjectNode config =
        new KafkaDependencyBuilder("kafka")
            .withPort(19092)
            .withContainerName("kafka-box")
            .withImageName("apache/kafka")
            .build()
            .forFfi();
    assertEquals(19092, config.path("port").asInt());
    assertEquals("kafka-box", config.path("container_name").asText());
    assertEquals("apache/kafka", config.path("image_name").asText());
  }

  @Test
  void addChildDependency_nonEmptyChildren_serializesChildrenArray() {
    ObjectNode config =
        new KafkaDependencyBuilder("kafka").addChildDependency(new StubDependency()).build().forFfi();
    assertEquals(1, ((ArrayNode) config.path("children")).size());
  }

  @Test
  void identifier_returnsConfiguredIdentifier() {
    KafkaDependency dep = new KafkaDependencyBuilder("kafka").build();
    assertTrue(dep.identifier().startsWith("arena-kafka-kafka-"));
  }
}
